File: LittleFencer-Android/app/src/main/java/com/littlefencer/app/logic/detectors/ParryRiposteDetector.kt
package com.littlefencer.app.logic.detectors

import com.littlefencer.app.logic.ActionDetector
import com.littlefencer.app.logic.ActionQuality
import com.littlefencer.app.logic.ActionResult
import com.littlefencer.app.logic.PoseFrame
import com.littlefencer.app.logic.SaberAction
import com.littlefencer.app.utils.GeometryUtils
import com.littlefencer.app.utils.MediaPipeLandmarks

/**
 * ParryRiposteDetector - Detects Parry (格挡) and Riposte (反攻) actions.
 * 
 * Saber Parry Positions:
 * - Tierce (3): Block to the outside, blade moves laterally
 * - Quarte (4): Block to the inside
 * - Quinte (5): Block to the head (high parry)
 * 
 * Biomechanics:
 * - Parry: Quick wrist movement to deflect, minimal arm movement
 * - Riposte: Immediate counter-attack following successful parry
 * 
 * Detection Strategy:
 * 1. Detect rapid wrist lateral/vertical movement (parry)
 * 2. Wrist returns to attack position quickly (riposte)
 * 3. Full parry-riposte should complete within ~500ms
 */
class ParryRiposteDetector : ActionDetector {

    override val targetAction = SaberAction.PARRY  // Primary action

    private var phase = Phase.IDLE
    private var phaseStartTime = 0L
    private var parryDirection = ParryDirection.NONE
    private var parryWristPos = 0f to 0f  // X, Y at parry peak
    
    private enum class Phase {
        IDLE,
        PARRYING,  // Defensive wrist movement
        RIPOSTING  // Counter-attack
    }
    
    private enum class ParryDirection {
        NONE,
        TIERCE,  // Outside (wrist moves away from body)
        QUARTE,  // Inside (wrist moves toward body)
        QUINTE   // High (wrist moves up for head protection)
    }

    companion object {
        // Parry detection
        private const val PARRY_LATERAL_THRESHOLD = 0.04f   // Min wrist lateral movement
        private const val PARRY_VERTICAL_THRESHOLD = 0.03f  // Min wrist vertical movement for quinte
        private const val PARRY_VELOCITY_THRESHOLD = 0.3f   // Quick defensive motion
        
        // Riposte detection
        private const val RIPOSTE_VELOCITY_THRESHOLD = 0.4f  // Forward velocity
        private const val MAX_PARRY_TO_RIPOSTE = 400L        // ms between parry and riposte
        
        // Timeouts
        private const val MAX_PARRY_DURATION = 300L
        private const val MAX_TOTAL_DURATION = 800L
    }

    override fun detect(current: PoseFrame, history: List<PoseFrame>): ActionResult {
        if (history.size < 5) return ActionResult.None

        val landmarks = current.landmarks
        if (landmarks.size < 33) return ActionResult.None

        // Determine front side and extract wrist positions
        val leftWrist = landmarks[MediaPipeLandmarks.LEFT_WRIST]
        val rightWrist = landmarks[MediaPipeLandmarks.RIGHT_WRIST]
        val leftShoulder = landmarks[MediaPipeLandmarks.LEFT_SHOULDER]
        val rightShoulder = landmarks[MediaPipeLandmarks.RIGHT_SHOULDER]
        
        val isLeftFront = leftWrist.x > rightWrist.x
        val frontWrist = if (isLeftFront) leftWrist else rightWrist
        val frontShoulder = if (isLeftFront) leftShoulder else rightShoulder
        
        // Get previous frame data
        val prevFrame = history.last()
        val prevLandmarks = prevFrame.landmarks
        if (prevLandmarks.size < 33) return ActionResult.None
        
        val prevFrontWrist = if (isLeftFront) {
            prevLandmarks[MediaPipeLandmarks.LEFT_WRIST]
        } else {
            prevLandmarks[MediaPipeLandmarks.RIGHT_WRIST]
        }
        
        val deltaTime = current.timeDeltaMs(prevFrame)
        
        // Calculate wrist velocities
        val lateralVelocity = if (deltaTime > 0) {
            (frontWrist.x - prevFrontWrist.x) / deltaTime * 1000f
        } else 0f
        
        val verticalVelocity = if (deltaTime > 0) {
            (frontWrist.y - prevFrontWrist.y) / deltaTime * 1000f
        } else 0f
        
        // Wrist position relative to shoulder
        val wristRelativeX = frontWrist.x - frontShoulder.x
        val wristRelativeY = frontWrist.y - frontShoulder.y

        return when (phase) {
            Phase.IDLE -> {
                // Detect parry initiation (rapid defensive wrist movement)
                val direction = detectParryDirection(lateralVelocity, verticalVelocity, isLeftFront)
                
                if (direction != ParryDirection.NONE) {
                    phase = Phase.PARRYING
                    phaseStartTime = current.timestampMs
                    parryDirection = direction
                    
                    ActionResult.InProgress(SaberAction.PARRY, 0.5f)
                } else {
                    ActionResult.None
                }
            }
            
            Phase.PARRYING -> {
                val elapsed = current.timestampMs - phaseStartTime
                
                // Check timeout
                if (elapsed > MAX_PARRY_DURATION) {
                    // Parry completed without riposte
                    val quality = evaluateParryQuality()
                    reset()
                    
                    return ActionResult.Completed(
                        action = SaberAction.PARRY,
                        quality = quality,
                        feedback = getParryName(),
                        durationMs = elapsed
                    )
                }
                
                // Detect riposte initiation (forward velocity after parry)
                if (lateralVelocity > RIPOSTE_VELOCITY_THRESHOLD && 
                    kotlin.math.abs(verticalVelocity) < PARRY_VELOCITY_THRESHOLD * 0.5f) {
                    
                    parryWristPos = frontWrist.x to frontWrist.y
                    phase = Phase.RIPOSTING
                    
                    ActionResult.InProgress(SaberAction.RIPOSTE, 0.7f)
                } else {
                    ActionResult.InProgress(SaberAction.PARRY, 0.6f)
                }
            }
            
            Phase.RIPOSTING -> {
                val totalElapsed = current.timestampMs - phaseStartTime
                
                // Check timeout
                if (totalElapsed > MAX_TOTAL_DURATION) {
                    reset()
                    return ActionResult.None
                }
                
                // Check completion (wrist extended forward past parry position)
                if (frontWrist.x > parryWristPos.first + 0.03f) {
                    val quality = evaluateRiposteQuality(lateralVelocity)
                    reset()
                    
                    ActionResult.Completed(
                        action = SaberAction.RIPOSTE,
                        quality = quality,
                        feedback = "${getParryName()} → Riposte!",
                        durationMs = totalElapsed
                    )
                } else {
                    ActionResult.InProgress(SaberAction.RIPOSTE, 0.8f)
                }
            }
        }
    }
    
    private fun detectParryDirection(lateralV: Float, verticalV: Float, isLeftFront: Boolean): ParryDirection {
        // Quinte (head parry) - upward movement
        if (verticalV < -PARRY_VELOCITY_THRESHOLD) {
            return ParryDirection.QUINTE
        }
        
        // Tierce/Quarte - lateral movement
        if (kotlin.math.abs(lateralV) > PARRY_VELOCITY_THRESHOLD) {
            return if ((lateralV < 0 && isLeftFront) || (lateralV > 0 && !isLeftFront)) {
                ParryDirection.TIERCE  // Outside
            } else {
                ParryDirection.QUARTE  // Inside
            }
        }
        
        return ParryDirection.NONE
    }
    
    private fun getParryName(): String = when (parryDirection) {
        ParryDirection.TIERCE -> "Tierce (外格)"
        ParryDirection.QUARTE -> "Quarte (内格)"
        ParryDirection.QUINTE -> "Quinte (头格)"
        ParryDirection.NONE -> "Parry"
    }
    
    private fun evaluateParryQuality(): ActionQuality {
        // Simple evaluation based on successful detection
        return ActionQuality.GOOD
    }
    
    private fun evaluateRiposteQuality(velocity: Float): ActionQuality {
        return when {
            velocity > RIPOSTE_VELOCITY_THRESHOLD * 1.5f -> ActionQuality.PERFECT
            velocity > RIPOSTE_VELOCITY_THRESHOLD -> ActionQuality.GOOD
            else -> ActionQuality.ACCEPTABLE
        }
    }

    override fun reset() {
        phase = Phase.IDLE
        phaseStartTime = 0L
        parryDirection = ParryDirection.NONE
        parryWristPos = 0f to 0f
    }
}

File: LittleFencer-iOS/LittleFencer/Logic/Detectors/ParryRiposteDetector.swift
import CoreGraphics
import Foundation

/// Detects Parry (格挡) and Riposte (反攻) actions
///
/// Saber Parry Positions:
/// - Tierce (3): Block to the outside, blade moves laterally
/// - Quarte (4): Block to the inside
/// - Quinte (5): Block to the head (high parry)
///
/// Biomechanics:
/// - Parry: Quick wrist movement to deflect, minimal arm movement
/// - Riposte: Immediate counter-attack following successful parry
class ParryRiposteDetector: ActionDetector {
    
    let targetAction = SaberAction.parry  // Primary action
    
    private enum Phase {
        case idle
        case parrying
        case riposting
    }
    
    private enum ParryDirection: String {
        case none = "None"
        case tierce = "Tierce"   // Outside
        case quarte = "Quarte"   // Inside
        case quinte = "Quinte"   // High
    }
    
    private var phase = Phase.idle
    private var phaseStartTime: Double = 0
    private var parryDirection = ParryDirection.none
    private var parryWristPos: (CGFloat, CGFloat) = (0, 0)
    
    // MARK: - Thresholds
    
    private let PARRY_LATERAL_THRESHOLD: CGFloat = 0.04
    private let PARRY_VERTICAL_THRESHOLD: CGFloat = 0.03
    private let PARRY_VELOCITY_THRESHOLD: CGFloat = 0.3
    private let RIPOSTE_VELOCITY_THRESHOLD: CGFloat = 0.4
    private let MAX_PARRY_TO_RIPOSTE: Double = 400
    private let MAX_PARRY_DURATION: Double = 300
    private let MAX_TOTAL_DURATION: Double = 800
    
    func detect(current: PoseFrame, history: [PoseFrame]) -> ActionResult {
        guard history.count >= 5, current.landmarks.count >= 33 else {
            return .none
        }
        
        let landmarks = current.landmarks
        
        // Extract wrist positions
        let leftWrist = landmarks[PoseLandmarks.LEFT_WRIST]
        let rightWrist = landmarks[PoseLandmarks.RIGHT_WRIST]
        let leftShoulder = landmarks[PoseLandmarks.LEFT_SHOULDER]
        let rightShoulder = landmarks[PoseLandmarks.RIGHT_SHOULDER]
        
        let isLeftFront = leftWrist.x > rightWrist.x
        let frontWrist = isLeftFront ? leftWrist : rightWrist
        let frontShoulder = isLeftFront ? leftShoulder : rightShoulder
        
        // Get previous frame data
        guard let prevFrame = history.last,
              prevFrame.landmarks.count >= 33 else {
            return .none
        }
        
        let prevFrontWrist = isLeftFront 
            ? prevFrame.landmarks[PoseLandmarks.LEFT_WRIST]
            : prevFrame.landmarks[PoseLandmarks.RIGHT_WRIST]
        
        let deltaTime = current.timeDeltaMs(prevFrame)
        
        // Calculate wrist velocities
        let lateralVelocity: CGFloat = deltaTime > 0 
            ? (frontWrist.x - prevFrontWrist.x) / CGFloat(deltaTime) * 1000 
            : 0
        let verticalVelocity: CGFloat = deltaTime > 0 
            ? (frontWrist.y - prevFrontWrist.y) / CGFloat(deltaTime) * 1000 
            : 0
        
        // Wrist position relative to shoulder
        let wristRelativeX = frontWrist.x - frontShoulder.x
        let wristRelativeY = frontWrist.y - frontShoulder.y
        
        switch phase {
        case .idle:
            // Detect parry initiation
            let direction = detectParryDirection(lateralVelocity: lateralVelocity,
                                                 verticalVelocity: verticalVelocity,
                                                 isLeftFront: isLeftFront)
            
            if direction != .none {
                phase = .parrying
                phaseStartTime = current.timestampMs
                parryDirection = direction
                return .inProgress(action: .parry, confidence: 0.5)
            }
            return .none
            
        case .parrying:
            let elapsed = current.timestampMs - phaseStartTime
            
            // Check timeout
            if elapsed > MAX_PARRY_DURATION {
                let quality = evaluateParryQuality()
                let feedback = getParryName()
                reset()
                return .completed(action: .parry, quality: quality,
                                 feedback: feedback, durationMs: elapsed)
            }
            
            // Detect riposte initiation
            if lateralVelocity > RIPOSTE_VELOCITY_THRESHOLD &&
               abs(verticalVelocity) < PARRY_VELOCITY_THRESHOLD * 0.5 {
                parryWristPos = (frontWrist.x, frontWrist.y)
                phase = .riposting
                return .inProgress(action: .riposte, confidence: 0.7)
            }
            
            return .inProgress(action: .parry, confidence: 0.6)
            
        case .riposting:
            let elapsed = current.timestampMs - phaseStartTime
            
            // Check timeout
            if elapsed > MAX_TOTAL_DURATION {
                reset()
                return .none
            }
            
            // Check riposte completion (forward velocity decreased, arm extended)
            if lateralVelocity < RIPOSTE_VELOCITY_THRESHOLD * 0.3 &&
               frontWrist.x > parryWristPos.0 + 0.05 {
                
                let quality = evaluateRiposteQuality(elapsed: elapsed)
                reset()
                return .completed(action: .riposte, quality: quality,
                                 feedback: "Nice riposte!", durationMs: elapsed)
            }
            
            return .inProgress(action: .riposte, confidence: 0.8)
        }
    }
    
    private func detectParryDirection(lateralVelocity: CGFloat, 
                                      verticalVelocity: CGFloat,
                                      isLeftFront: Bool) -> ParryDirection {
        // Quinte - high parry (upward movement)
        if verticalVelocity < -PARRY_VERTICAL_THRESHOLD &&
           abs(lateralVelocity) < PARRY_VELOCITY_THRESHOLD {
            return .quinte
        }
        
        // Tierce - outside parry (wrist moves away from body)
        if isLeftFront && lateralVelocity > PARRY_LATERAL_THRESHOLD {
            return .tierce
        } else if !isLeftFront && lateralVelocity < -PARRY_LATERAL_THRESHOLD {
            return .tierce
        }
        
        // Quarte - inside parry (wrist moves toward body)
        if isLeftFront && lateralVelocity < -PARRY_LATERAL_THRESHOLD {
            return .quarte
        } else if !isLeftFront && lateralVelocity > PARRY_LATERAL_THRESHOLD {
            return .quarte
        }
        
        return .none
    }
    
    private func evaluateParryQuality() -> ActionQuality {
        // Simplified quality evaluation
        switch parryDirection {
        case .quinte: return .good  // High parry is harder
        case .tierce, .quarte: return .good
        case .none: return .acceptable
        }
    }
    
    private func evaluateRiposteQuality(elapsed: Double) -> ActionQuality {
        // Quick riposte is better
        if elapsed < 300 { return .perfect }
        if elapsed < 500 { return .good }
        return .acceptable
    }
    
    private func getParryName() -> String {
        return "Good \(parryDirection.rawValue)!"
    }
    
    func reset() {
        phase = .idle
        phaseStartTime = 0
        parryDirection = .none
        parryWristPos = (0, 0)
    }
}
